articles/051_actions.md

```diff
@@ -0,0 +1,95 @@
+author: '[Geoffrey Biggs](https://github.com/gbiggs)'
+published: true
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+ROS services, which provide synchronous Remote Procedure Calls, are a useful concept for sending a request and getting a rapid reply. But in robotics there are many instances where a reply may take a significant length of time. Additionally, there are occasions when it is useful to send a request to do some processing or perform some action in the world, where the result is less important than the effect of carrying it out. The progress of such requests often needs to be tracked, success or failure must be known in addition to receiving back information produced, and the request may need to be cancelled or altered before it completes. These requirements cannot be fulfilled by a simple RPC mechanism, whether or not it is asynchronous.
+
+To satisfy these use cases, ROS provides a third communication paradigm known as "actions". An action is a goal-oriented request that occurs asynchronously to the requester, is typically (but not necessarily) longer-running than immediate, can be cancelled or replaced during execution, and has a server that provides feedback on execution progress.
+
+This document defines how actions are specified, what they look like to ROS users (both node developers and system integrators)
+
+## Action specification
+
+Actions are specified using a form of the ROS Message IDL. The specification contains three sections, each of which is a message specification:
+
+1. Goal
+
+1. Result
+
+1. Feedback
+
+Any of these sections may be empty.
+
+Between the three sections is a line containing three hyphens, `---`.
+
+Action specifications are stored in a file ending in `.action`. There is one action specification per `.action` file.
+
+An example action specification [taken from the actionlib wiki] is shown below.
+
+```
+# Define the goal
+uint32 dishwasher_id  # Specify which dishwasher we want to use
+---
+# Define the result
+uint32 total_dishes_cleaned
+---
+# Define a feedback message
+float32 percent_complete
+uint32 number_dishes_cleaned
+```
+
+## Serving and using actions
+
+Actions are a first-class citizen in the ROS API, alongside topics and services.
+
+Action clients will use an API that provides a proxy object for the action. This will be a templated class, using the action class generated from the action specification as the template parameter. The client shall create an instance of this class, providing the address of the intended action server. Each instance of this class can only be related to one action server. Methods of the class will provide facilities for sending a goal to the action server, receiving a result, and getting feedback.
```

Comment:
In order to enforce the one action server per address, we would need to ensure this at the service level which we don't right now. This is tied to the node name uniqueness issue which we still haven't addressed yet.

We can set that as a goal, but when it comes time to implement we may just have to push this problem for the time being.